keps/sig-storage/2605-option-to-preserve-default-file-mode-bit-in-atomicwriter-volumes/README.md

```diff
@@ -0,0 +1,181 @@
+    // If unspecified, each key-value pair in the Data field of the referenced
+    // Secret will be projected into the volume as a file whose name is the
+    // key and content is the value. If specified, the listed keys will be
+    // projected into the specified paths, and unlisted keys will not be
+    // present. If a key is specified which is not present in the Secret,
+    // the volume setup will error unless it is marked optional. Paths must be
+    // relative and may not contain the '..' path or start with '..'.
+    // +optional
+    Items []KeyToPath
+    // Mode bits to use on created files by default. Must be a value between
+    // 0 and 0777.
+    // Directories within the path are not affected by this setting.
+    // This might be in conflict with other options that affect the file
+    // mode, like fsGroup, and the result can be other mode bits set.
+    // If PreservePermissions is set to true then file mode bit is not
+    // modified after the file creation, due to any other options set
+    // such as fsGroup etc
+    // +optional
+    DefaultMode *int32
+    // Specify whether the Secret or its key must be defined
+    // +optional
+    Optional *bool
+    // Specify whether we want to preserve the DefaultMode set on the files
+    // at the time of file creation.
+    // If PreservePermissions is set to true the file mode is not modified based
+    // on fsGroup after the file is created with the defaultMode.
+    // If the PreservePermissions is set to false and fsGroup is specified
+    // the volume is modified to be owned by the fsGroup.
+    // If PreservePermissions is not specified in the volume spec, it defaults to false.
+    // +optional
+    PreservePermissions *bool
+}
+```
+
+A secret volume source with PreservePermissions=true:
+
+```yaml
+  volumes:
+  - name: test-secret
+    secret:
+      preservePermissions: true
+      defaultMode: 256
+      secretName: test-secret
+```
+
+### File Permission
+
+For the volumes that use AtomicWriter today the files are created with the Mode set in the KeyToPath struct.
+If the Mode is not specified then the DefaultMode set for the source is used.
+If DefaultMode is not specified then the file is created with mode bit set to 0600.
+If pod have fsGroup specified then all the files in the volume are walked,
+chown and chmod to the fsGroup
+
+#### Proposed heuristics
+
+- *Case 1*: The volume has PreservePermissions set to false or not specified.
+    In this case there is no behavioural change in how the file permissions are
+    set for the volume. In other words, we will continue to create the files
+    with Mode/Default and then set the volume ownership based on fsGroup if specified.
+    This is the default case.
+- *Case 2*: The volume has PreservePermissions set to true.
+    In this case the payload/volume/files are created using Mode/Default mode.
+    Then the call to SetVolumeOwnership is skipped, in turn preserving the Default file mode on the files.
```

Comment:
Have we tried commenting out call to `SetVolumeOwnership` and just setting the mode on files? It sounds like this - this will cause files within container to become inaccessible to the container's process.  I am not sure if setting the mode is enough.